### ISSUE TITLE
DeepSeek-V3 Model Status Tracking

### ISSUE THREAD

**Issue by wwwjn** [pytorch/torchtitan#1727]
## Model Functionality
- [Done] Parallelism
    - FSDP, TP, PP, EP, ETP
    - Activation checkpointing
- [Done] Numerical verification (comparing against HuggingFace model) 
- [In progress] StateDictAdapter (conversion between HuggingFace) 
    - There are 2 important issues to solve: 1) Dequantization, because the 671B model weights on HF are quantized. 2) Individual <-> Grouped Experts weights conversion, because torchtitan and HF has different MoE experts definition. 
    - [In progress] Dequantization: Solved by @saumishr  by designing `QuantizedHuggingFaceStorageReader`. (https://github.com/pytorch/pytorch/pull/160682/files) 
    - [In progress] Individual <-> Grouped Experts weights conversion: Designed and implemented new algorithm
    - [In progress] end-to-end numerical verification 

## BenchMarking 
- We verified loss convergence on deepseek-v3 671b model:  
     - This run is trained from scratch
- We launched jobs to benchmarking DeepSeek-v3 671B performance (this might be a little bit legacy and it was done in August) 
    - Currently DeepSeek-v3 671B model are runnable on 256 GPUs (H100).
    - Example Parallelism settings on 90GB H100:
         - FSDP=32, TP=8, local bs=8; 
         - FSDP=32, TP=8, EP=8, local bs=4. 

## What are we working on?
We have some known performance issue:
- Incompatibility between FSDP and Full AC, causing memory issue #1624  (PoC: @fegin @weifengpy  @soulitzer )
- DualPipe-style overlapping: #1682 (PoC: @H-Huang )
- DeepEP-style communication dedup (PoC: @kwen2501 )

Other issue related to DeepSeek-model:
- SAC + FlexAttention + compile: #1683  (PoC: @fegin ). 

Feature enablement: 
- FP8 Training for deepseek-V3 @danielvegamyhre 
- CP support for deepseek-v3

Maybe correctness issue
- PP microbatching + block-causal FlexAttention composability issue: #1723

**Comment by rakkit** [comment]
hi@wwwjn, will EP+Compile(`fullgraph`) be considered in the roadmap?  If not, do we consider enabling full graph compile for the attention block?

**Comment by tianyu-l** [comment]
Hey @rakkit thanks for asking!

Since FSDP2 is graph breaking, and we do `compile(TransformerBlock)` and `FSDP(moe.experts)`, so we will have graph break around `moe.experts` anyway.

Currently we are working on get a full graph within experts + a2a region. @xmfan can share more on progress.

May I ask what you'd care most about when you mention full graph compile + EP / Attention?

**Comment by rakkit** [comment]
Thanks for the update, @tianyu-l  main reason is for memory efficiency and a slight throughput improvement.
This is because, at (`fullgraph=False`) case, the compiler is not smart enough to fuse lots OPs, which is suprisely bad.

Here is one of the examples back to our test on Torch 2.6, for `TransformerBlock` ([here](https://github.com/rakkit/torchtitan/blob/trustllm-0.1/torchtitan/models/MoEllama/GroupedExperts.py#L112) and [here](https://github.com/rakkit/torchtitan/blob/trustllm-0.1/torchtitan/models/MoEllama/moellama.py#L855))
```
    @torch.compile(fullgraph=True)
    def _apply_norm(self, norm, x):
        return norm(x)

    def forward(x, freqs_cis):
      h = x + self.attention(self._apply_norm(self.attention_norm, x), freqs_cis)

      mlp_output, _ = self.feed_forward(self._apply_norm(self.ffn_norm, h))
```

manually forcing `full_graph=true` compile for all `norms` can give a significant memory saving.

So basically, we just manually add lots of `@torch.compile(fullgraph=True)` in MoE and whole attention, and we get a significant memory saving and a bit of efficiency improvement. 


And for the current version, at least we could have the full graph for attention? 

And we could have a full graph for non-EP case, e.g. in our branch, we do
```
def apply_compile(model: nn.Module, ep_enabled: bool):
    """
    Apply torch.compile to each TransformerBlock, which makes compilation efficient due to
    repeated structure. Alternatively one can compile the whole model (after applying DP).
    """

    for layer_id, transformer_block in model.layers.named_children():
        # we dont do compile when EP enabled
        fullgraph = True and not ep_enabled
        transformer_block = torch.compile(transformer_block, fullgraph=fullgraph)
        model.layers.register_module(layer_id, transformer_block)
```

**Comment by tianyu-l** [comment]
@rakkit 
Even if you don't get fullgraph on `TransformerBlock` but get several partial graphs, let's say the graph break is from MoE / EP, you should still get all `Attention` covered in the same partial graph (and therefore the memory savings). And the rest should run as fast as eager.

If something worse happens in the following case, we should investigate
> This is because, at (`fullgraph=False`) case, the compiler is not smart enough to fuse lots OPs, which is suprisely bad.

Btw, are you using FlexAttn with SAC? We were recently working on https://github.com/pytorch/torchtitan/issues/1631

**Comment by rakkit** [comment]
Thanks a lot for the explanation @tianyu-l, I will come back later with some benchmarks on the latest repo and torch, see if we can have any speedup via manually wrapping.

(I am not use sac + flexattn)
